packages/profile-sync-controller/src/controllers/user-storage/network-syncing/sync-all.ts

```diff
@@ -0,0 +1,238 @@
+import { setDifference, setIntersection } from '../utils';
+import {
+  toRemoteNetworkConfiguration,
+  type NetworkConfiguration,
+  type RemoteNetworkConfiguration,
+  toNetworkConfiguration,
+} from './types';
+
+type FindNetworksToUpdateProps = {
+  localNetworks: NetworkConfiguration[];
+  remoteNetworks: RemoteNetworkConfiguration[];
+};
+
+const createMap = <
+  Network extends NetworkConfiguration | RemoteNetworkConfiguration,
+>(
+  networks: Network[],
+): Map<string, Network> => {
+  return new Map(networks.map((n) => [n.chainId, n]));
+};
+
+const createKeySet = <
+  Network extends NetworkConfiguration | RemoteNetworkConfiguration,
+>(
+  networks: Network[],
+  predicate?: (n: Network) => boolean,
+): Set<string> => {
+  const filteredNetworks = predicate
+    ? networks.filter((n) => predicate(n))
+    : networks;
+  return new Set(filteredNetworks.map((n) => n.chainId));
+};
+
+export const getDataStructures = (
+  localNetworks: NetworkConfiguration[],
+  remoteNetworks: RemoteNetworkConfiguration[],
+) => {
+  const localMap = createMap(localNetworks);
+  const remoteMap = createMap(remoteNetworks);
+  const localKeySet = createKeySet(localNetworks);
+  const remoteKeySet = createKeySet(remoteNetworks);
+  const existingRemoteKeySet = createKeySet(remoteNetworks, (n) => !n.d);
+
+  return {
+    localMap,
+    remoteMap,
+    localKeySet,
+    remoteKeySet,
+    existingRemoteKeySet,
+  };
+};
+
+type MatrixResult = 'Do Nothing' | 'Local Wins' | 'Remote Wins';
```

Comment:
See [figma](https://www.figma.com/board/fZTsDekl5xbHo8IheI3MMc/Network-Syncing-Notes?node-id=0-1&node-type=canvas&t=tWpXvIVguaMgdpYG-0).

This is the main result matrix that intersecting networks are compared against.